### PR TITLE
feat(assistant): fill the analysis tracker from the catalog and harden output parsing

### PR DESCRIPTION
**Pull request by dannon**
The Analysis Assistant builds an "Analysis Setup" tracker as the conversation
goes, and only hands off to workflow setup once it's complete. Today that
tracker is filled entirely from the SCHEMA_UPDATE lines the model emits -- fine
with a strong model, but it broke in two ways I kept hitting against smaller,
self-hosted models (the TACC MiniMax / gpt-oss endpoints):

- Fields that aren't user choices never filled. "Data characteristics"
  (paired/single-end reads, or an assembled genome) and "Gene annotation" are
  properties of the chosen workflow and assembly, not user decisions -- so the
  model correctly never emits a SCHEMA_UPDATE for them, and they sat Pending,
  blocking the handoff. I derive them deterministically from the catalog
  instead: data characteristics from the workflow's read inputs (or a FASTA
  genome input for annotation workflows), gene annotation from the selected
  assembly's gene model. Same "the catalog decides, don't make the model guess"
  principle the rest of the platform follows.

- The model dropped SCHEMA_UPDATE lines. Smaller models intermittently skip the
  line on the turn a decision is made, and the prompt only asked for it then --
  one missed turn left the tracker permanently behind. The prompt now asks the
  model to re-emit every committed field each turn, which is self-healing.

Separately, I saw raw SUGGESTIONS/SCHEMA_UPDATE JSON leak into the visible reply
when the model emitted it inline, split across lines, or with mangled brackets
-- the old parser only matched a marker at line start and required valid JSON.
The parser now scans the whole reply, decodes the JSON wherever it sits, and
excises an uppercase marker's payload even when it can't parse (so raw JSON is
never shown), while leaving lowercase prose like "a few suggestions: [...]"
alone.

Gene-annotation reconciliation also auto-fills the GTF when the assembly ships
one instead of flagging it for attention (#1324/#1331).

Five commits, each green on its own; tests cover the catalog-derived fills and
the parser's inline / multi-line / mangled-JSON / prose cases.

### CODE REVIEW

**Comment by NoopDog** [comment]
Reviewed this with Claude Code — solid hardening, and the happy-path test coverage is thorough. Two issues I'd flag before merge, both reproduced by running this branch's code:

**1. A malformed `SCHEMA_UPDATE` trailer silently drops the `SUGGESTIONS` that follow it (`assistant_agent.py` ~L904).**
On the `JSONDecodeError` path, `last = max(text.rfind("]"), text.rfind("}"))` scans the whole reply, so the splice runs from the marker to the last bracket *anywhere after it*. The prompt asks for `SUGGESTIONS` "at the very end," after `SCHEMA_UPDATE` — so in the normal trailer order, a malformed `SCHEMA_UPDATE` reaches through and deletes the valid `SUGGESTIONS` block. Reproduced on this branch:
```
"I found a few variant-calling workflows that fit your data.
SCHEMA_UPDATE: {"organism": "Plasmodium falciparum", "workflow": broken
SUGGESTIONS: ["Variant calling", "Transcriptomics"]"
```
→ prose is kept, but chips come back empty. `test_malformed_suggestions_before_schema_update_preserved` guards the reverse order; this is the order the prompt actually requests, and the MiniMax sample in the tests shows malformed JSON does occur. Suggest bounding the excision to the marker's own value region rather than `rfind` across the whole reply, plus a regression test for this ordering.

**2. Data-characteristics derivation misses the `_FILE`/`FORWARD_FILE`/`REVERSE_FILE` read variants (`_data_characteristics_for_workflow`, ~L1132).**
It matches only `SANGER_READ_RUN_PAIRED`/`SANGER_READ_RUN_SINGLE`, but active catalog workflows use the file variants as their read input — "Genome assembly with Flye" (`SANGER_READ_RUN_SINGLE_FILE`) and "Bacterial Genome Assembly using Shovill" (`FORWARD_FILE` + `REVERSE_FILE`). Running the real method against both returns `None`, so `data_characteristics` is set Empty and the tracker stays incomplete — the same bug this PR is fixing. The `library_layout` is already in each param's `data_requirements`, so it's just a matter of recognizing those variable names (forward/reverse ⇒ paired).

**Comment by dannon** [comment]
Thanks @NoopDog -- both good catches, and both are fixed on the branch:

- **Malformed `SCHEMA_UPDATE` eating the following `SUGGESTIONS`:** reproduced your exact case. The recovery splice was reaching to the last bracket anywhere in the reply; I bounded it to the marker's own payload (stops at the next real trailer), with a regression test for that ordering.
- **`_FILE`/`FORWARD_FILE`/`REVERSE_FILE` read variants:** hardened the derivation to match the whole `SANGER_READ_RUN_*` family and use the declared `library_layout`. One nuance -- those workflows (Flye, Shovill) are organism-scope, so they don't reach the derivation through the assistant today, but this keeps it correct if they ever go assembly-scope.

Bigger picture: chasing these (plus a Codex adversarial pass) made clear that most of this fragility is inherent to scraping structured state back out of the model's prose. **Structural changes are coming in subsequent PR(s) to make this more robust** -- moving the schema/suggestion channel to native structured output so this whole class of parsing edge-cases stops existing, rather than getting patched one at a time. This PR still lands the reconciliation hardening + property/fuzz tests; the scrape replacement will be its own focused PR with a model eval.

Thanks again for the thorough look.

**Comment by dannon** [comment]
Wrapping this one up. This round hardened the `SCHEMA_UPDATE`/`SUGGESTIONS` parsing and tracker reconciliation about as far as it's worth taking -- two Codex adversarial passes, Dave's review, and several Copilot rounds, all addressed (threads resolved) -- plus a hypothesis property net that catches the parser class automatically now.

Most of these were symptoms of one root cause: machine state riding in the model's prose and getting regex'd back out. We've already been scoping the real fix as the follow-up -- schema/suggestions to typed, schema-validated tool calls with pydantic-ai in-loop validation/retry -- and #1429 lines up with that direction.

Where the version we're building is deliberately more conservative is on weaker models. The whole reason #1408 exists is the self-hosted gpt-oss / MiniMax class, and pydantic-ai's in-loop retry only helps once the model *calls* the tool -- it does nothing if a small model doesn't reliably emit tool calls at all, in which case a typed-only channel gets nothing and the tracker stalls, which is worse than today. So we're treating "remove the parser fallback" as an eval call per model rather than a given -- keeping a per-model fallback where the eval says we need one, plus a crossover-leak guard for models that both tool-call and still emit a legacy line. We'll drive the follow-up with those baked in.

Treating this PR as ready to merge on its own merits -- deterministic catalog fills + a more robust parser -- with the structural work continuing in the follow-up.

File: backend/api/app/services/assistant_agent.py [hunk 1]
```diff
@@ -832,53 +847,137 @@ async def restore_saved_session(
...
+                    # A real trailer is either at line start (the model's own
+                    # line) or written with some uppercase -- the markers are
+                    # emitted in caps per the prompt. An all-lowercase marker
+                    # mid-prose (e.g. "...a few suggestions: [...]") is prose and
+                    # is left untouched, even when the brackets are valid JSON.
+                    line_start = text.rfind("\n", 0, m.start()) + 1
+                    at_line_start = text[line_start : m.start()].strip() == ""
+                    marker_is_real = at_line_start or any(
+                        c.isupper() for c in m.group(0)
+                    )
```
[the remark this answers is not in the document]
Comment (dannon) [reply backend/api/app/services/assistant_agent.py:900]: Tabling this for the structured-output rework (#1429) -- a typed `propose_suggestions` tool turns a wrong-type payload like this into a validation error rather than a fallback-parser leak, so it gets handled there instead of with more heuristics on a channel we're replacing.

**Comment by NoopDog** [comment]
**Notes from a Claude-assisted code review** 🤖

I took a careful pass through this (parser state machine + the reflector chain) and ran the actual `_parse_structured_output` / walked the reflectors against the catalog to verify. The direction is great — catalog-authoritative derivation and whole-reply marker scanning are the right calls, and the leak-prevention holds up on the common cases. A few things surfaced that seemed worth flagging, roughly in severity order. All are FYI for your judgment — nothing here is a hard blocker from my side.

**1. An assembly-scope workflow can never complete/hand off.**
`is_complete()` requires `data_characteristics` (`models/assistant.py:59`), and `_reflect_data_characteristics` now force-sets it EMPTY whenever `_data_characteristics_for_workflow` returns `None` (no `SANGER_READ_RUN*` / `ASSEMBLY_FASTA_URL` param). There's a real one in the catalog: **`consensus-peaks-atac-cutandrun`** (`scope: ASSEMBLY`, `parameters: []`). Selecting it fills organism/assembly/analysis_type/workflow/data_source, but `data_characteristics` derives to EMPTY, so `is_complete()` stays `False` forever and the handoff never appears — with no recovery, since the reflector re-derives EMPTY on every apply (the `if not updates: return current` early-return being removed means it runs each turn). It slightly defeats the PR's own goal for the no-declared-input case. Might be worth treating `data_characteristics` as satisfied (or not-required) when a FILLED assembly-scope workflow has nothing to derive. The derivation test asserts EMPTY but doesn't check `is_complete`, so the interaction isn't currently covered.

**2. The parser occasionally discards legitimate content (reproduced by running the method).**
Prevents raw-JSON leaks well, but a few excision cases drop valid text:

| Input | Output | Issue |
|---|---|---|
| `Recorded.\nSCHEMA_UPDATE: {"a": broken\nSUGGESTIONS:\n[...]` | visible `'Recorded.'`, suggestions `[]` | valid next-line SUGGESTIONS swallowed by the malformed SCHEMA_UPDATE |
| `SCHEMA_UPDATE: {oops\nThanks for choosing us...` | visible `''` | over-excision to end-of-text → empty reply |
| `...organism SCHEMA_UPDATE: {"organism":"X"} and pick an assembly next.` | visible `'...organism'` | inline trailer drops the rest of the sentence |

Root cause of the first two: the boundary scan at `assistant_agent.py:923` does `after = text[bm.end():].lstrip(" \t:*")` then checks `after[:1] in ("[","{")` — but `lstrip` doesn't include `\n`, so a following trailer whose JSON opens on the next line (a layout supported elsewhere and covered by `test_suggestions_multiline_json`) isn't seen as a boundary, and the excision runs to the end. Adding `\n` to that lstrip set looks like it'd resolve the swallowed-SUGGESTIONS and most of the over-excision. These are the small-model malformed outputs the rewrite is aimed at, so they feel in-scope.

**3. Sentence-capitalized marker mid-prose creates spurious chips.**
`marker_is_real = at_line_start or any(c.isupper() ...)` fires on ordinary capitalization: `Here are my thoughts. Suggestions: ["read docs","ask forum"] should help.` → visible `'Here are my thoughts.'` + two chips, and "should help." is dropped. The all-caps xfail tests don't cover the single-leading-capital case.

**4. A prompt example contradicts the new instruction.**
The SCHEMA_UPDATE section now asks to re-emit **all** committed fields every turn, but the "switch from RNA-seq to variant calling" example (`assistant_agent.py:278`) emits only the changed field + nulls, dropping the already-committed `organism`/`assembly`. Since few-shot examples steer strongly, restating all committed fields in that example would reinforce rather than undercut the new behavior.

**5. gene_annotation NEEDS_ATTENTION vs handoff (pre-existing, just noting).**
When a workflow needs a GTF and the assembly has none, `gene_annotation` is NEEDS_ATTENTION, but `is_complete()` omits `gene_annotation`, so handoff still fires. Not introduced here (`is_complete` is unchanged), but the reflector rewrite sits right on it — worth a deliberate decision on whether that should block.

**Test-quality FYI (non-blocking):** the hypothesis strategies draw from `a–z` + space, so generated JSON never contains `"{}[]:` or marker words — the string-literal-bracket case the comments call "tricky" is exercised only by the hand-written examples, not the property net. And the single-end **name-inference** branch isn't hit (the test also supplies `library_layout`, short-circuiting it).

Really nice work overall — the derivation model is a clear improvement over model-emitted state. Happy to pair on any of these if useful.

**Review by NoopDog — APPROVED** [review]

**Comment by dannon** [comment]
Thanks @NoopDog. Disposition on each, mostly folded into #1430 (the stacked extraction-pass PR on the same code):

1. **Assembly-scope workflow can't complete** -- real bug, fixed. A resolved workflow with no derivable read/FASTA input now reports "Provided at workflow setup" so `is_complete()`/handoff aren't blocked forever. Kept the wording neutral on purpose: a codex pass pointed out the catalog build drops `type_guide`-only params (your `consensus-peaks` example actually needs a BAM), so it doesn't claim "no input required."
2. **Parser over-excision** -- confirmed. Tried your next-line-boundary fix, but a codex pass showed it opens a leak when a marker sits inside a broken JSON string, so I kept the conservative over-excise-to-end (no-leak wins) and documented it as an accepted limitation.
3. **Capitalized marker mid-prose** -- real gap; accepted limitation (can't cleanly tell a sentence-cap "Suggestions:" from a trailer). Lower-impact now the reply is plain text.
4. **Prompt example contradiction** -- moot in #1430: the reply no longer emits `SCHEMA_UPDATE` (state rides a separate extractor), so that section is gone.
5. **gene_annotation NEEDS_ATTENTION vs handoff** -- pre-existing, unchanged; flagging for a product call on whether it should block.

Also fixed a related gap this surfaced: restore computed handoff off persisted state without reconciling, so a session on a since-removed workflow could hand off on stale detail -- it now reconciles first.
